### PR TITLE
[DTensor] assume uneven sharding by default, redistribute cost hints for unbacked

### PR DESCRIPTION
**Pull request by pianpwk**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #175287
* #174629
* __->__ #168052

This extends compiling unbacked matmuls to strategies that require redistribution.

- When negative/zero cost strategies aren't available, cost expressions are functions of shapes, and so are unbacked, and there's no obvious minimum cost choice. The PR follows the inductor size-hinting strategy of using a fallback value, first checking for optimization hints, then using the symbol upper-bound (as a proxy for worst-case redistribution costs), then falling back to a hardcoded value (8192).
- Avoids using `torch.chunk()` in collective implementations, opting for more unbacked-friendly implementations. Chunk is DDE-prone because it calls `split(tensor, chunk_size)` under the hood, which loses track of the original # of chunks, and tries to extract a concrete value from an ugly unbacked expression. In these cases we typically know the # of chunks beforehand (group size), so this seems avoidable.
- Without divisibility axioms for unbacked sizes, assumes that we should handle uneven sharding by default, so pad/narrow ops are run before/after collectives to be divisible by group size (even if the runtime value runs 0 padding or no-op slices). Users can add `torch._check(size % num_shards)` to opt into the faster path.

Fixes 32 DDE-ing DTensor ops

With this PR, going through all 256 input placements (on R/S0/S1/P) of a matmul on a 2d mesh traces through (eager/aot_eager), following this sample script: https://gist.github.com/pianpwk/65a7e3a60354c8dcdbd532b597d96826

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

### CODE REVIEW

File: torch/distributed/tensor/_sharding_prop.py [hunk 1]
```diff
@@ -581,6 +582,88 @@ def propagate_op_sharding_non_cached(self, op_schema: OpSchema) -> OutputShardin
...
+        """
+        from torch.fx.experimental.symbolic_shapes import (
+            free_unbacked_symbols,
+            is_concrete_float,
+        )
+        from torch.utils._sympy.interp import sympy_interp
+        from torch.utils._sympy.numbers import int_oo
+        from torch.utils._sympy.reference import PythonReferenceAnalysis
+
+        int_fallback = 8192
```
Comment (pianpwk) [inline torch/distributed/tensor/_sharding_prop.py]: wasn't sure if there was a config to put this in

File: torch/distributed/tensor/_sharding_prop.py [hunk 2]
```diff
@@ -581,6 +582,88 @@ def propagate_op_sharding_non_cached(self, op_schema: OpSchema) -> OutputShardin
...
+
+        # Use replacements for redistribute cost hints
+        proxy_costs = [
+            float(cost)
+            if is_concrete_float(cost)
+            else sympy_interp(
+                PythonReferenceAnalysis,
+                replacements,
+                cost.node.expr.xreplace(replacements),  # type: ignore[arg-type]
+            )
```
Comment (pianpwk) [inline torch/distributed/tensor/_sharding_prop.py]: subs/xreplace wouldn't simplify to a concrete value, so this seemed like the way?

File: c10/core/ConstantSymNodeImpl.h [hunk 3]
```diff
@@ -68,6 +68,8 @@ class C10_API ConstantSymNodeImpl : public SymNodeImpl {
   c10::SymNode lt(const c10::SymNode& other) override;
   c10::SymNode gt(const c10::SymNode& other) override;
   c10::SymNode mul(const c10::SymNode& other) override;
+  c10::SymNode sym_and(const c10::SymNode& other) override;
+  c10::SymNode sym_or(const c10::SymNode& other) override;
```
Comment (ezyang) [inline c10/core/ConstantSymNodeImpl.h:72]: This could be split out into its own PR, if the other changes here get stuck in limbo

File: torch/distributed/_functional_collectives.py [hunk 4]
```diff
@@ -208,7 +208,11 @@ def all_gather_tensor(
         # and then chunk + cat avoid us going through ACT dispatching logic again
         if isinstance(res, AsyncCollectiveTensor):
             res = res.wait()  # type: ignore[attr-defined]
-        res = torch.cat(torch.chunk(res, group_size, dim=0), dim=gather_dim)
+        chunk_size = res.size(0) // group_size
+        chunks = [
+            torch.narrow(res, 0, i * chunk_size, chunk_size) for i in range(group_size)
+        ]
+        res = torch.cat(chunks, dim=gather_dim)
```
Comment (ezyang) [inline torch/distributed/_functional_collectives.py]: This may negatively affect eager mode performance; you may be obligated to only do this when doing compile tracing.

File: torch/distributed/_local_tensor/__init__.py [hunk 5]
```diff
@@ -543,6 +553,11 @@ def ge(self, other: "int | LocalIntNode | ConstantIntNode") -> bool | SymBool:
         assert len(r) == 1, (self, other)
         return torch._C._get_constant_bool_symnode(next(iter(r)))
 
+    def le(self, other: "int | LocalIntNode | ConstantIntNode") -> bool | SymBool:
+        r = {self._local_ints[r] <= _int_on_rank(other, r) for r in self._local_ints}
+        assert len(r) == 1, (self, other)
+        return torch._C._get_constant_bool_symnode(next(iter(r)))
```
Comment (ezyang) [inline torch/distributed/_local_tensor/__init__.py:557]: Ah OK, these LocalTensor changes look fine

File: torch/distributed/tensor/_sharding_prop.py [hunk 6]
```diff
@@ -581,6 +582,88 @@ def propagate_op_sharding_non_cached(self, op_schema: OpSchema) -> OutputShardin
...
+                    for spec in strat.input_specs
+                )
+            placements_out = tree_map(
+                lambda spec: spec.format_shard_order_str(
+                    spec.placements, spec.shard_order
+                ),
+                strat.output_specs,
+                is_leaf=lambda x: isinstance(x, DTensorSpec),
+            )
+            log.info(
```
Comment (ezyang) [inline torch/distributed/tensor/_sharding_prop.py]: This seems too spammy. It should be debug level, IMO. Also, it's not clear to me this log is actually usable outside of if you're literally just running a DTensor prop rule.

File: torch/distributed/tensor/_sharding_prop.py [hunk 7]
```diff
@@ -581,6 +582,88 @@ def propagate_op_sharding_non_cached(self, op_schema: OpSchema) -> OutputShardin
...
+                    spec.format_shard_order_str(spec.placements, spec.shard_order)
+                    for spec in strat.input_specs
+                )
+            placements_out = tree_map(
+                lambda spec: spec.format_shard_order_str(
+                    spec.placements, spec.shard_order
+                ),
+                strat.output_specs,
+                is_leaf=lambda x: isinstance(x, DTensorSpec),
+            )
```
Comment (ezyang) [inline torch/distributed/tensor/_sharding_prop.py]: This logic here not occurring inside of a lambda means that you are always doing the computation for this (which is expensive! you're string formatting) EVEN if the log message doesn't show up anywhere.

File: torch/distributed/tensor/placement_types.py [hunk 8]
```diff
@@ -129,6 +124,35 @@ def _select_split_tensor(
...
+                tensor_list, dim, num_chunks - len(tensor_list)
+            )
+        else:
+            dim_size = tensor.size(dim)
+            split_size = (dim_size + num_chunks - 1) // num_chunks
+            chunks = []
+            for i in range(num_chunks):
+                start = torch.sym_min(split_size * i, dim_size)
+                end = torch.sym_min(split_size * (i + 1), dim_size)
+                chunks.append(tensor.narrow(dim, start, end - start))
```
Comment (wconstab) [inline torch/distributed/tensor/placement_types.py:204]: concerned this would semantically differ from torch.chunk in terms of how many chunks it created when the mesh dim size (num_chunks) is > the tensor dim size.  (e.g. `torch.chunk(torch.ones((1,)), chunks=10)` returns
`(tensor([1.]),)`)

Comment (pianpwk) [reply torch/distributed/tensor/placement_types.py:204]: ah, I think that's what the previous `fill_empty_tensor_to_shards` was for

Comment (wconstab) [reply torch/distributed/tensor/placement_types.py:204]: yea. but is it ok that you don't have that in the tracing branch? i would have thought no

Comment (pianpwk) [reply torch/distributed/tensor/placement_types.py:204]: I think this loop slicing should already produce the empty tensors...

Comment (pianpwk) [reply torch/distributed/tensor/placement_types.py:204]: yeahh, in the previous impl, chunk would produce k < num_chunks entries, then call `fill_empty_tensor_to_shards` to fill the num_chunks - k rest.

This impl should just have num_chunks calls to narrow, with the later ones being empty slices

File: torch/distributed/tensor/_collective_utils.py [hunk 9]
```diff
@@ -173,7 +173,9 @@ def mesh_broadcast(
 
 @maybe_run_for_local_tensor
 def pad_tensor(tensor: torch.Tensor, pad_dim: int, pad_size: int) -> torch.Tensor:
-    if pad_size == 0:
+    from torch.fx.experimental.symbolic_shapes import guard_or_false
+
+    if guard_or_false(pad_size == 0):
         return tensor
```
Comment (wconstab) [inline torch/distributed/tensor/_collective_utils.py:181]: cc @IvanKobzarev in the case you mentioned where some ranks got padding and others didn't- 
do you know if you're using unbacked symints for tracing here, or backed?

it matters becuase @IvanKobzarev was wanting the property that he always gets padding ops in his graph even if the pad size is 0, so we may want to argue for this condition only applying when not tracing, or just get rid of this condition altogether

File: torch/distributed/tensor/placement_types.py [hunk 10]
```diff
@@ -182,8 +206,11 @@ def local_shard_size_and_offset(
         Returns (new local shard size, offset)
 
         """
+        from torch.fx.experimental.symbolic_shapes import guard_or_false
```
Comment (wconstab) [inline torch/distributed/tensor/placement_types.py:225]: how come these imports are not done at the top, is it actually a circular import?

File: torch/distributed/tensor/placement_types.py [hunk 11]
```diff
@@ -194,15 +221,11 @@ def local_shard_size_and_offset(
...
-        if curr_local_size < shard_starting_idx:
-            # pyrefly: ignore[bad-return]
-            return 0, typing.cast(_RankTypeT, curr_local_size)
-        else:
-            local_shard_size = (
-                sym_min(curr_local_size, shard_starting_idx + full_chunk_size)
-                - shard_starting_idx
-            )
-            return local_shard_size, shard_starting_idx
+        shard_end_idx = torch.sym_min(
```
Comment (wconstab) [inline torch/distributed/tensor/placement_types.py:240]: just curious, how come we have sym_min instead of plumbed symints through min? does this just dispatch to min in the non-symint case?

File: torch/distributed/_functional_collectives.py [hunk 12]
```diff
@@ -99,6 +99,24 @@ def is_torchdynamo_compiling():  # type: ignore[misc]
 ]
 
 
+def _chunk_or_narrow_cat(
```
Comment (wconstab) [inline torch/distributed/_functional_collectives.py]: how come we don't use this helper in the change above in _redistribute.py? it seems almost the same but i'm not sure the difference is intentional

File: torch/distributed/tensor/_collective_utils.py (hunk 9 above)
Comment (pianpwk) [reply torch/distributed/tensor/_collective_utils.py:181]: fwiw, changing to `statically_known_true(pad_size == 0)` should default to padding in the backed case too

File: torch/distributed/tensor/placement_types.py (hunk 10 above)
Comment (pianpwk) [reply torch/distributed/tensor/placement_types.py:225]: there's a `test_not_import_sympy` that avoids importing anything from symbolic_shapes when we first import torch. We end up having to do this quite a lot actually

File: torch/distributed/tensor/placement_types.py (hunk 11 above)
Comment (pianpwk) [reply torch/distributed/tensor/placement_types.py:240]: yeah, we should just evaluate to the min value for non-symints, and keep a `sym_min(a, b)` expression in the symbolic case

File: torch/distributed/_functional_collectives.py (hunk 12 above)
Comment (pianpwk) [reply torch/distributed/_functional_collectives.py]: do you mean `_custom_chunk` ? I guess in this case we already know divisibility by number of chunks beforehand (they should be pre-padded)?

File: torch/distributed/tensor/placement_types.py (hunk 10 above)
Comment (wconstab) [reply torch/distributed/tensor/placement_types.py:225]: ok

File: torch/distributed/_functional_collectives.py (hunk 12 above)
Comment (wconstab) [reply torch/distributed/_functional_collectives.py]: the one i was referring to was shard_dim_all_to_all_meta

```
    cat_tensor = torch.cat(stacked_list, dim=gather_dim)
    # pyrefly: ignore [unsupported-operation]
    chunk_size = cat_tensor.size(shard_dim) // group_size
    chunk = torch.narrow(cat_tensor, shard_dim, group_rank * chunk_size, chunk_size)
    return chunk.contiguous()
```
though i realized now it is not the same thing.

File: torch/distributed/tensor/_collective_utils.py (hunk 9 above)
Comment (wconstab) [reply torch/distributed/tensor/_collective_utils.py:181]: can you explain what statically_known_true(pad_size==0) does?

Comment (pianpwk) [reply torch/distributed/tensor/_collective_utils.py:181]: It's kinda like guard_or_false for backed symints; unless we can prove the thing statically (no peeking at hints), we'll return False.

I'm not sure how to best expose this as an option though, if the case is really with backed symints here

Comment (wconstab) [reply torch/distributed/tensor/_collective_utils.py:181]: sounds like if we gaurantee that the same static constraints were used on all ranks then using statically_known_true is safe even for ivan's case

**Review by wconstab — APPROVED** [review]

File: torch/_utils.py [hunk 13]
```diff
@@ -1176,9 +1176,12 @@ def _maybe_view_chunk_cat(
     # Optimization: Can use view instead of split+cat when:
     # 1. res.shape[0] == group_size (invariant after all_gather)
     # 2. All dims between 0 and gather_dim (exclusive) have size 1
+    from torch.distributed._functional_collectives import _chunk_or_narrow_cat
+    from torch.fx.experimental.symbolic_shapes import guard_or_false
+
     numel_between = math.prod(shape[1:gather_dim]) if gather_dim > 1 else 1
 
-    if shape[0] == group_size and numel_between == 1:
+    if guard_or_false(shape[0] == group_size) and numel_between == 1:
```
Comment (laithsakka) [inline torch/_utils.py]: guard_or_false around numel_between == 1:?

File: torch/distributed/tensor/_sharding_prop.py [hunk 14]
```diff
@@ -135,6 +137,98 @@ def cache_clear(self):
...
+    """
+    from torch.fx.experimental.symbolic_shapes import (
+        free_unbacked_symbols,
+        is_concrete_float,
+    )
+    from torch.utils._sympy.interp import sympy_interp
+    from torch.utils._sympy.numbers import int_oo
+    from torch.utils._sympy.reference import PythonReferenceAnalysis
+
+    int_fallback = 8192
```
Comment (laithsakka) [inline torch/distributed/tensor/_sharding_prop.py:186]: This is fine(ish) can we add a BIG TODO that this shall be rewritten to use the optimization_hint API that we will add in symbolic_shapes and link a github issue you can assign it to me i want to track that this is not like this forever

**Comment by laithsakka** [comment]
looks fine, lets just create issue to track unification of hinting code.

File: torch/distributed/tensor/_sharding_prop.py (hunk 14 above)
Comment (pianpwk) [reply torch/distributed/tensor/_sharding_prop.py:186]: added a todo in your name haha

File: torch/distributed/tensor/_sharding_prop.py [hunk 15]
```diff
@@ -135,6 +137,99 @@ def cache_clear(self):
...
+    free_unbacked = list(set(chain(*[free_unbacked_symbols(cost) for cost in costs])))
+
+    # Easy path: no unbacked shapes involved, choose min cost strategy.
+    # Doing the hard path for backed could also make sense?
+    if all(is_concrete_float(c) for c in costs) or not free_unbacked:
```
Comment (zpcore) [inline torch/distributed/tensor/_sharding_prop.py:191]: Do we need to check both? Is `not free_unbacked` sufficient?

File: test/distributed/tensor/test_dtensor_compile.py [hunk 16]
```diff
@@ -603,6 +629,54 @@ def test_dtensor_matmul_zero_size_shards(self):
...
+        )
+        y_dt = DTensor.from_local(
+            torch.randn(8, 8),
+            device_mesh=device_mesh,
+            placements=[Shard(1)],
+        )
+        torch._dynamo.decorators.mark_unbacked(x_dt, 0)
+        torch._dynamo.decorators.mark_unbacked(y_dt, 1)
+
+        # large k dim tells compiler it's cheaper to all-gather on x
```
Comment (zpcore) [inline test/distributed/tensor/test_dtensor_compile.py:703]: k dim means y.size(1)?

**Review by zpcore — APPROVED** [review]
